de.fu_berlin.inf.dpp.intellij/src/de/fu_berlin/inf/dpp/intellij/project/SharedResourcesManager.java

```diff
@@ -195,126 +195,194 @@ private void handleFileRecovery(FileActivity activity) throws IOException {
 
-        //if the file exists, try to replace the content completely
-        if (!newText.isEmpty()) {
-            //this is true only when the file already existed
-            boolean replaceSuccessful = localEditorManipulator
-                .replaceText(activity.getPath(), newText);
+            return;
+        }
 
-            if (replaceSuccessful) {
-                //If the content of the existing document was replaced
-                //successfully, save the file
-                localEditorHandler.saveDocument(activity.getPath());
-                return;
-            }
+        InputStream contents = new ByteArrayInputStream(activity.getContent());
+
+        try {
+            fileSystemListener.setEnabled(false);
+
+            file.create(contents, FORCE);
+
+        } finally {
+            fileSystemListener.setEnabled(true);
         }
+    }
+
+    private void handleFolderCreation(
+        @NotNull
+            FolderCreatedActivity activity) throws IOException {
 
-        //If the file did not exist, create it
-        IFile file = activity.getPath().getFile();
-        byte[] actualContent = FileUtils.getLocalFileContent(file);
-        byte[] newContent = activity.getContent();
+        IFolder folder = activity.getPath().getFolder();
+
+        if (folder.exists()) {
+            LOG.warn(
+                "Could not create folder " + folder + " as it already exist.");
 
-        if (Arrays.equals(newContent, actualContent)) {
-            LOG.info("FileActivity " + activity + " dropped (same content)");
             return;
         }
 
         try {
             fileSystemListener.setEnabled(false);
-            FileUtils.writeFile(new ByteArrayInputStream(newContent), file);
-            //HACK: It does not work to disable the fileSystemListener temporarily,
-            //because a fileCreated event will be fired asynchronously,
-            //so we have to add this file to the filter list
-            fileSystemListener
-                .addIncomingFileToFilterFor(file.getLocation().toFile());
+
+            folder.create(FORCE, LOCAL);
+
         } finally {
             fileSystemListener.setEnabled(true);
         }
     }
 
-    private void handleFolderActivity(IFileSystemModificationActivity activity)
-        throws IOException {
+    /**
+     * Applies the given FolderDeletedActivity.
+     * <p>
+     * <b>NOTE:</b> This currently does not check whether the deleted folder
+     * contains resources outside the session scope. As a result, submodules of
+     * the shared module that are not present for a different participant can be
+     * deleted accidentally through remote activities.
+     * </p>
+     *
+     * @param activity the FolderDeletedActivity to execute
+     * @throws IOException if the folder deletion fails
+     */
+    //TODO deal with children that are not part of the current session (submodules)
+    private void handleFolderDeletion(
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Move this method into the anonymous class declared at line 105. ([squid:S3398](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS3398))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=277&issues=AWagwjc6lHK86HcoyTJE&open=AWagwjc6lHK86HcoyTJE)

[ISSUE_KEY]: AWagwjc6lHK86HcoyTJE